src/Castle.Core/DynamicProxy/Generators/AttributesToAvoidReplicating.cs

```diff
@@ -41,18 +42,11 @@ static AttributesToAvoidReplicating()
 
 		public static void Add(Type attribute)
 		{
-			// notes:
-			// this class is made thread-safe by replacing the backing list rather than adding to it
-			// this loop runs until the attributes collection contains the new attribute
-			// it is necessary to ensure that if multiple threads concurrently replace the list that they
-			// don't override each others changes
-			IList<Type> originalAttributes;
-			while (!(originalAttributes = attributes).Contains(attribute))
+			// note: this class is made thread-safe by replacing the backing list rather than adding to it
+			lock (lockObject)
```

Comment:
I think you also need to put `lock`s around places where `attributes` is accessed for reading; that is, in `Contains` and in `ShouldAvoid`. (Not sure about the former, but almost certain in case of the latter.)

Reply:
IMO not necessary. Exchanging object references is atomic. So `Contains` and `ShouldAvoid` will either get the old reference or the new reference.
Bug #334 is fixed by just replacing the list instead of extending it. However, the lock is necessary to safeguard against concurrent calls to `Add`. Even though they are not very likely, of course.

Reply:
Oops, I mis-read the diff (or rather, I didn't read all of it). You're absolutely right, of course.